database/buildSchema/43_views_functions_triggers.sql

```diff
@@ -1516,6 +1516,74 @@ GROUP BY
+-- this becomes less then trivial because when allowed_sections is null it means that all sections can be
+-- reviewed by the user, thus we need to deduce how many sections there are total in template and replace
+-- allowed sections with all sections in template (when allowed sections is null)
+CREATE OR REPLACE FUNCTION public.allowed_self_assignable_sections (userId int)
+      RETURNS SETOF allowed_self_assignable_sections_shape
+    AS $$
+   SELECT 
+    review_assignment.id as review_assignment_id,
+    CASE WHEN allowed_sections IS NULL THEN
+        array_agg(distinct(template_section.code))
+    ELSE
+        allowed_sections
+    END AS allowed_sections
+    FROM review_assignment
+        JOIN application ON review_assignment.application_id = application.id
+        JOIN template ON application.template_id = template.id
+        JOIN template_section ON template_section.template_id = template.id
+    WHERE review_assignment.status = 'AVAILABLE' 
+        AND review_assignment.is_self_assignable = TRUE
+        AND application.outcome = 'PENDING'
+        AND review_assignment.reviewer_id = userId
+    -- Need to group by allowed sections because it appears in 'from' clause 
+    -- (in CASE statement), there would be just one allowed_sections field
+    -- for reach group by, so it should be safe
+    GROUP BY 1, review_assignment.allowed_sections;
+$$
+LANGUAGE sql
+STABLE;
+
+CREATE TABLE IF NOT EXISTS review_assignment_assigned_sections_shape (
+    review_assignment_id int,
+    assigned_sections varchar[]
+);
+
+-- In order to deduce if there are any self assignable sections for we need to know which sections 
+-- are already assigned, which requires looking at all 'ASSIGNED' review assignments and concatinating
+-- array of all sections in those assignments, looks like unnesting and array_agg is the only way to do this ?
+CREATE OR REPLACE FUNCTION public.review_assignment_assigned_sections (userId int)
+      RETURNS SETOF review_assignment_assigned_sections_shape
+    -- This allows the function to be run as admin user, otherwise it'll only be
+    -- able to query rows the current user has access to.
+    SECURITY DEFINER
+    AS $$
+    SELECT
+        ra_outer.id AS review_assignment_id,
+        ARRAY_AGG(DISTINCT(ra_inner.section_code))
+    FROM review_assignment AS ra_outer
+    JOIN application on ra_outer.application_id = application.id
+    LEFT JOIN (select *, unnest(assigned_sections) AS section_code FROM review_assignment) AS ra_inner 
+        ON (ra_inner.application_id = ra_outer.application_id
+            AND ra_inner.stage_id = ra_outer.stage_id 
+            AND ra_inner.level_number = ra_outer.level_number
+            AND ra_inner.status = 'ASSIGNED')
+    WHERE ra_outer.status = 'AVAILABLE' 
+        AND ra_outer.is_self_assignable = TRUE
+        AND application.outcome = 'PENDING'
+        AND ra_outer.reviewer_id = userId
+    GROUP BY 1
+$$
+LANGUAGE sql
+STABLE;
```

Comment:
We already do something very similar to this with the `enforce_asssigned_section_validity` function (near the top of this script). I can't help thinking it would be good to just add another couple of columns to the review_assignment table where we can store the result of this calculation too. The trigger is already in place to update it whenever it changes, and it's been pretty robust, so I don't think we'd have a problem keeping it up to date.

Let's discuss...

Reply:
This would be helpful, but it would only work for subset of reviewer actions, although I think it would still speed it up a bit (since these are the slow ones)

Reply:
Yeah, I'll keep thinking about this, try and document all the relevant events/actions and see if it makes sense.